devito/finite_differences/derivative.py

```diff
@@ -0,0 +1,162 @@
+import sympy
+
+from collections import OrderedDict
+from devito.finite_differences.finite_difference import (generic_derivative,
+                                                         first_derivative,
+                                                         cross_derivative)
+from devito.finite_differences.differentiable import Differentiable
+from devito.finite_differences.tools import centered, direct, transpose, left, right
+from devito.tools import as_tuple
+
+
+class Derivative(sympy.Derivative, Differentiable):
+
+    """
+    An unevaluated Derivative, which carries metadata (Dimensions,
+    derivative order, etc) to be evaluated.
+
+    Parameters
+    ----------
+
+    expr : symbolic expression
+    dims : Dimension or tuple
+        Dimenions wrt which to take the derivative
+    deriv_order: Int or Tuple
+        Order of the derivative for each Dimension
+    fd_order : Int or tuple
+        Order of the finite difference for each Dimension
+    stagger : Dimension or tuple
+        Staggering for each dimension
+    side : Side or tuple
+        Side of the finite difference for each Dimension
+    transpose : direct or adjoint
+        Wether the finite difference is transposed
+
+    Examples
+    --------
+    Creation
+
+    >> from devito import Function, Derivative, Grid
+    >> grid = Grid((10, 10))
+    >> u = Function(name="u", grid=grid, space_order=2)
+    >> Derivative(u, u.indices[0])
+    Derivative(u(x, y), x)
+
+    This can also be obtained via the differential shortcut
+
+    >> u.dx
+
+    For higher order you can specify the order as a keyword argument
+
+    >> Derivative(u, x, deriv_order=2)
+    Derivative(u(x, y), (x, 2))
+
+    Or as a tuple
+
+    >> Derivative(u, (x, 2))
+    Derivative(u(x, y), (x, 2))
+
+    Once again, this can be access via the differential shortcut
+
+    >> u.dx2
+    """
+
+    def __new__(cls, expr, *dims, **kwargs):
+        # Check dims, can be a dimensions, multiple dimensions as a tuple
+        # or a tuple of tuple (ie ((x,1),))
+        if len(dims) == 1:
+            if isinstance(dims[0], (tuple, sympy.Tuple)):
```

Comment:
what if one provides a list? or more generically, an `Iterable` ?

I **think** you want:

```
from collections import Iterable  # quite sure it lives in collections, otherwise find out
if isinstance(dims[0], Iterable)
```